server/sonar-server/src/main/java/org/sonar/server/computation/step/DuplicationMeasuresStep.java

```diff
@@ -59,63 +67,149 @@
   private final MetricRepository metricRepository;
   private final MeasureRepository measureRepository;
 
-  public DuplicationMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository) {
+  public DuplicationMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository,
+    @Nullable DuplicationRepository duplicationRepository) {
     this.treeRootHolder = treeRootHolder;
     this.metricRepository = metricRepository;
     this.measureRepository = measureRepository;
-    this.formulas = ImmutableList.<Formula>of(
-      new SumDuplicationFormula(DUPLICATED_BLOCKS_KEY),
-      new SumDuplicationFormula(DUPLICATED_FILES_KEY),
-      new DuplicationFormula());
+    this.formulas = ImmutableList.<Formula>of(new DuplicationFormula(metricRepository, measureRepository, duplicationRepository));
+  }
+
+  /**
+   * Constructor used by Pico in Views where no DuplicationRepository is available.
+   */
+  public DuplicationMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this(treeRootHolder, metricRepository, measureRepository, null);
+  }
+
+  @Override
+  public String getDescription() {
+    return "Compute duplication measures";
   }
 
   @Override
   public void execute() {
     new PathAwareCrawler<>(
       FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository).buildFor(formulas))
-        .visit(treeRootHolder.getRoot());
+      .visit(treeRootHolder.getRoot());
   }
 
-  private class DuplicationFormula implements Formula<SumDuplicationCounter> {
+  private static class DuplicationCounter implements Counter<DuplicationCounter> {
+    @CheckForNull
+    private final DuplicationRepository duplicationRepository;
+    private int fileCount = 0;
+    private int blockCount = 0;
+    private int lineCount = 0;
+
+    private DuplicationCounter(@Nullable DuplicationRepository duplicationRepository) {
+      this.duplicationRepository = duplicationRepository;
+    }
+
+    @Override
+    public void aggregate(DuplicationCounter counter) {
+      this.fileCount += counter.fileCount;
+      this.blockCount += counter.blockCount;
+      this.lineCount += counter.lineCount;
+    }
+
+    @Override
+    public void initialize(CounterInitializationContext context) {
+      Component leaf = context.getLeaf();
+      if (leaf.getType() == Component.Type.FILE) {
+        initializeForFile(leaf);
+      } else if (leaf.getType() == Component.Type.PROJECT_VIEW) {
+        initializeForProjectView(context);
+      }
+    }
+
+    private void initializeForFile(Component file) {
+      DuplicationRepository duplicationRepository = requireNonNull(this.duplicationRepository, "DuplicationRepository missing");
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename "duplicationRepository" which hides the field declared at line 98. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)